howso/hyperparameters.amlg

```diff
@@ -913,4 +965,393 @@
+		(declare (assoc guessing_accuracy_matrix (call !ClampSDMProbabilities) ))
+
+		;compute the full deviation matrix
+		(declare (assoc
+			deviation_matrix
+				(map
+					(lambda (let
+						(assoc
+							predicted_class (current_index 1)
+							;probability of correct predicted class
+							class_probability (get (current_value 1) (current_index 1))
+							;total probability of the row
+							total_probability (apply "+" (values (current_value 1)) )
+						)
+
+						;normalize each prediction probability by iterating over all the predictions
+						;in the row and dividing by total_probability
+						(map
+							(lambda
+								(- 1 (/ (min class_probability (current_value)) total_probability))
+							)
+							(current_value)
+						)
+					))
+					guessing_accuracy_matrix
+				)
+		))
+
+		;store default class deviations for any classes that are not in the sparse confusion matrix
+		(declare (assoc
+			leftover_class_deviation_map
+				(filter (map
+					(lambda
+						;only need default if a class isn't in the sparse confusion matrix
+						(if (< (size (current_value)) num_classes)
+							(first (values
+								(remove (get deviation_matrix (current_index)) (indices (current_value)))
+							))
+						)
+					)
+					confusion_matrix
+				))
+		))
+
+		;if there's a leftover_unknown_class_deviation, output the SDM as a list (pair)
+		(if leftover_unknown_class_deviation
+			(list (call !FormatSDMForOutput) leftover_unknown_class_deviation)
+
+			;else just output the SDM as an assoc
+			(call !FormatSDMForOutput)
+		)
+	)
+
+	;Edits deviation_matrix for output by filtering out deviations that correspond to counts that aren't in the sparse confusion matrix
+	;and return tuples with the default class deviation for classes that have those computed
+	#!FormatSDMForOutput
+	(map
+		(lambda
+			;if there is a default deviation for the class, output a tuple of [sparse deviation map, default class deviation]
+			(if (contains_index leftover_class_deviation_map (current_index))
+				(list
+					(keep (current_value 1) (indices (get confusion_matrix (current_index 1))) )
+					(get leftover_class_deviation_map (current_index 1))
+				)
+
+				;just output the sdm assoc
+				(keep (current_value) (indices (get confusion_matrix (current_index))) )
+			)
+		)
+		deviation_matrix
+	)
+
+	;Helper method for computing SDM, filters out counts below threshold and stores as probabilities correct instead.
+	;leaves only those predicted classes that had enough predictions.
+	#!ConvertAndReduceConfusionMatrix
+	(seq
+		(assign (assoc
+			leftover_confusion_matrix
+				(map
+					(lambda (let
+						(assoc
+							predicted_class (current_index 1)
+							row_map (current_value 1)
+						)
+						;leave entire row out if predicted class doesn't have enough samples
+						(if
+							(or
+								(= (null) (get row_map predicted_class))
+								(< (get row_map predicted_class) smallest_count_threshold)
+							)
+							row_map
```

Comment:
if the second condition passes, we return the whole row, but the comment suggests we should leave the row out. Should this return null instead?